src/reuse/__init__.py

```diff
@@ -114,13 +109,56 @@ class ReuseInfo:
     source_path: Optional[str] = None
     source_type: Optional[SourceType] = None
 
+    def _check_nonexistent(self, **kwargs) -> None:
+        nonexistent_attributes = set(kwargs) - set(self.__dict__)
+        if nonexistent_attributes:
+            raise KeyError(
+                f"The following attributes do not exist in"
+                f" {self.__class__}: {', '.join(nonexistent_attributes)}"
+            )
+
+    def copy(self, **kwargs) -> Type["ReuseInfo"]:
+        """Return a copy of ReuseInfo, replacing the values of attributes with
+        the values from *kwargs*.
+        """
+        self._check_nonexistent(**kwargs)
+        new_kwargs = {}
+        for key, value in self.__dict__.items():
+            new_kwargs[key] = kwargs.get(key, value)
+        return self.__class__(**new_kwargs)
+
+    def union(self, value) -> Type["ReuseInfo"]:
+        """Return a new instance of ReuseInfo where all Set attributes are equal
+        to the union of the set in *self* and the set in *value*.
+
+        All non-Set attributes are set to their values in *self*.
+
+        >>> one = ReuseInfo(copyright_lines={"Jane Doe"}, source_path="foo.py")
+        >>> two = ReuseInfo(copyright_lines={"John Doe"}, source_path="bar.py")
+        >>> result = one.union(two)
+        >>> print(sorted(result.copyright_lines))
+        ['Jane Doe', 'John Doe']
+        >>> print(result.source_path)
+        foo.py
+        """
+        new_kwargs = {}
+        for key, attr_val in self.__dict__.items():
```

Comment:
Just checking: there is no need to check the keys of the other dictionary as all keys are added on initiation and so the structure is the same.

Reply:
Correct, provided you have two `ReuseInfo` objects.

I considered the cases `ReuseInfo.copy(SuperReuseInfo)` and `SuperReuseInfo.copy(ReuseInfo)`, but didn't write tests for them.

In the first case, the extra fields in `SuperReuseInfo` are not copied over into the new object, because they're not in `ReuseInfo`'s __dict__.

In the second case, the extra fields in `SuperReuseInfo` _cannot_ be overwritten because they do not exist in `ReuseInfo`.